### ISSUE TITLE
systemctl kexec appears to be non-functional / too strict requirements

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#15029]
It seems I can't get `systemctl kexec` to work.
I first tried it on a Debian based system where it simply yields
```
# systemctl kexec
No boot loader entry suitable as default, refusing to guess.
```
This system uses UEFI but does not follow the boot loader spec.

Let's try a Fedora 32 VM :
```
# systemctl kexec
No kexec kernel loaded and autodetection failed.
Automatic loading works only on systems booted with EFI.
```
Ah crap, I forgot to switch from BIOS to UEFI. Let's try again with a F32/UEFI VM:
```
# systemctl kexec
No boot loader entry suitable as default, refusing to guess.
```
No luck either.

Is `systemctl kexec` actually working for anyone?
Why the restriction to limit this to UEFI based systems and requiring the boot loader spec?
How can I make this work for a Debian (Bios) based system?

**Comment by mbiebl** [comment]
Related https://alioth-lists.debian.net/pipermail/pkg-systemd-maintainers/2020-February/040573.html

**Comment by cdown** [comment]
It works fine on non-UEFI for me, as long as you configure kexec properly first:

    # kexec -l /boot/vmlinuz-linux --initrd=/boot/initramfs-linux.img --append="cgroup_no_v1=all root=... rw"
    # systemctl kexec

**Comment by mbiebl** [comment]
That's not very userfriendly though. Ideally `systemctl kexec` should just DTRT.

**Comment by cdown** [comment]
Doing the right thing requires that we can reliably determine what the right thing is, or at least detect when we don't know what the right thing to do is. How do you intend to determine that with an innumerable number of bootloaders, boot configurations, and boot methods (eg. PXE, UEFI special sauce, etc)?

**Comment by poettering** [comment]
We DTRT, problem is how to find the kernels though. We implement the boot loader spec for finding the kernels. If you boot loader or its scripts don't support that this is not going to work though.

Anyway, there's nothing we are going to to do about this. Basically you have three options:

1. Explicitly, manually load the kernel you later want to kexec into via "kexec" first.
2. Use a boot loader that either supports the boot loader spec natively
3. Convince your distro that at least generate the boot loader specs as side effect in /run whenever they regenerate their boot loader config, if the boot loader doesn't support it natively, so that we can consume them.

And no, we are not going to parse grub's config or so to find kernels to boot. And not do distro-specific stuff either.

hence closing.

**Comment by mbiebl** [comment]
Well, it appears even for a distro which seemingly supports the boot  loader spec, like Fedora, `systemctl kexec` does not seem to work. This is issue number one.

Issue number two is the bad user experience. The error messages are very cryptic and not very user friendly.
If `systemctl exec` had output something like in https://github.com/systemd/systemd/issues/15029#issuecomment-596158645
telling me to manually load the kernel manually, this would have helped a lot.

Third, those restrictions make `systemctl kexec` basically useless for the majority of systems out there. I can understand that you don't want to parse grub boot loader entries yourself, but there is also no interface which would allow a 3rd party to hook into `systemctl kexec` and provide that information.

**Comment by mbiebl** [comment]
If there was such an integration point, I could e.g. ship a shell snippet for grub2 in Debian, which provides that information.

**Comment by poettering** [comment]
Yeah, we could make the messages more explanatory.

> I can understand that you don't want to parse grub boot loader entries yourself, but there is also no interface which would allow a 3rd party to hook into systemctl kexec and provide that information.

As mentioned: we do support that: it's the boot loader spec. The spec just defines a format for boot loader entries, it's extremely trivial, and it's supposed to be usable as a common format between packaging tools, boot loaders and systemd to talk about boot entries. If the boot loader of your choice doesn't support it, then that's a pity, but it's still a great way how your packaging scripts can communicate kernel info to systemd. i.e. systemd will pick up the data from some dir in /run too if the the boot loader doesn#t speak things natively, we are fine with that.

So, we *do* have that integration point. A shell script in the grub2 package could trivially generate the snippets as well, and we'd pick them up then. The integration point is the same for everyone: the boot loader spec. It's very lose coupling, i.e. just a common location where people put stuff, and it's good that way.

**Comment by mbiebl** [comment]
> As mentioned: we do support that: it's the boot loader spec. The spec just defines a format for boot loader entries, it's extremely trivial, and it's supposed to be usable as a common format between packaging tools, boot loaders and systemd to talk about boot entries

On a fedora 31 / UEFI  system:

```
# systemctl kexec
No boot loader entry suitable as default, refusing to guess.

# find /boot/loader/
/boot/loader/
/boot/loader/entries
/boot/loader/entries/d4225952fa354616b0aabf5a2a7bc308-5.5.11-200.fc31.x86_64.conf
/boot/loader/entries/d4225952fa354616b0aabf5a2a7bc308-0-rescue.conf
```

Repeating my question above: Is `systemctl kexec` actually working for anyone?

**Comment by poettering** [comment]
grub on fedora took the boot loader spec and turned it into a monster. It's nuts. Ignore that. They turned it into a macro language with pattern expansion and so on. It's totally crazy. And it really sucks they just took our name and turned it into such a monster.

Install sd-boot on fedora and all is good. it's packaged, but unfortunately not the default.

**Comment by mbiebl** [comment]
Which basically confirms my point that `systemctl kexec` is basically non-functional for the overwhelming majority out there.

**Comment by poettering** [comment]
Well, for people who upload the kernel with the "kexec" tools themselves things are fine. Exactly like they always were. The boot loader spec stuff is to make this more automatic, but it only works if people buy into a common spec. I mean, there is no other spec, no other interface than the boot loader spec, hence it's the one and only thing we support for this.

But this is no regression, because the boot loader spec makes things automatic where it can, and before it there was nothing, it was all manual.

(Arch uses systemd-boot by default iirc, it all just works for them)

Debian has two options: just make the kernel install scripts spit out boot loader spec entries (possibly in addition, parallel to whatever else they generate) and thus make the automatic stuff in systemd work, if they want "systemd kexec" work. or they can package kexec-tools in a way so that a kernel is always uploaded before "systemctl kexec" does its thing.

**Comment by mbiebl** [comment]
> Well, for people who upload the kernel with the "kexec" tools themselves things are fine. Exactly like they always were. The boot loader spec stuff is to make this more automatic, but it only works if people buy into a common spec. I mean, there is no other spec, no other interface than the boot loader spec, hence it's the one and only thing we support for this.
> 
> But this is no regression, because the boot loader spec makes things automatic where it can, and before it there was nothing, it was all manual.

There definitely is a regression behaviour wise. With older systemd versions, systemd would start kexec.target and the kexec-tools package in Debian had a script which did load the kernel and everything just worked.
Nowadays `systemctl exec` bails out early not every attempting to start kexec.target.

> (Arch uses systemd-boot by default iirc, it all just works for them)
> 
> Debian has two options: just make the kernel install scripts spit out boot loader spec entries (possibly in addition, parallel to whatever else they generate) and thus make the automatic stuff in systemd work, if they want "systemd kexec" work. or they can package kexec-tools in a way so that a kernel is always uploaded before "systemctl kexec" does its thing.

Always uploading the kernel in advance seems wasteful.

**Comment by hugelgupf** [comment]
> Install sd-boot on fedora and all is good. it's packaged, but unfortunately not the default.

Hey, I ran into this problem and found this thread. I just installed a vanilla Fedora 32 server edition, and there is no `sd-boot` or `systemd-boot` package to install. Is there some other name you meant to refer to?

I noticed that the `bootctl` tool did exist, provided by the `systemd-udev` package, and the following steps gave me an uncontaminated BLS config:

```shell
dnf remove grub2*
bootctl install
rm -rf /boot/loader # do this so bootctl won't assume these entries are already there, when they're in the wrong partition
kernel-install add $(uname -r) /lib/modules/$(uname -r)/vmlinuz
```

Is there a package that would automate this part, especially the kernel-install? Or is there a way to install Fedora with a pure systemd-boot in the first place? (I assume that after this, kernel updates already call kernel-install in some kind of post-install script, so that's taken care of -- just need to automate this on first install?)

If I should take my question over to Fedora, let me know...

**Comment by felagund** [comment]
On Ubuntu 20.10, when I set it up to use systemd-boot, `systemctl kexec` works. But there does not seem to be any obvious way to make it the default reboot action (i.e. clicling on reboot in Gnome Shell would initiate kexec). Is that a systemd issue or distribution issue?

**Comment by npmccallum** [comment]
I stumbled on this issue while having my own struggles with `systemd kexec`. The context for this is the [Wyrcan](https://gitlab.com/wyrcan/wyrcan) project which downloads a docker container image containing a kernel, extract its all and boots it. This will never have an entry in the bootloader spec because it isn't an installed system.

In my case I'm loading the kexec kernel using [kexec_file_load](https://manpages.debian.org/testing/manpages-dev/kexec_file_load.2.en.html) directly (no `kexec-tools`). Now I want to perform an orderly shutdown into this new kernel. Except it isn't obvious how to do this.

`systemctl kexec` seems to imply that it does an orderly shutdown followed by `reboot(LINUX_REBOOT_CMD_KEXEC)`. None of the documentation that I read (for example, `man systemctl` or `systemctl --help`) outlines that this command will *additionally* load a kernel from the bootloader spec before starting the orderly shutdown. This causes quite a bit of confusion as I have successfully loaded my own ephemeral kernel but when I do a `systemctl kexec` I silently get the system's kernel rather than the ephemeral one I just loaded.

So what is the right method to do an orderly shutdown and `reboot(kexec)` with a pre-loaded kernel?

**Comment by medhefgo** [comment]
> Is that a systemd issue or distribution issue?

You would need to convince gnome/plasma people to call `RebootWithFlags(1)` on logind instead of the regular one for this.

> So what is the right method to do an orderly shutdown and `reboot(kexec)` with a pre-loaded kernel?

systemd makes sure to check if a kexec kernel is already loaded before loading one itself, so `systemctl kexec` is the proper way. Maybe your systemd is too old? You could check what is happening with `SYSTEMD_LOG_LEVEL=debug systemctl --dry-run kexec`.

**Comment by npmccallum** [comment]
> > Is that a systemd issue or distribution issue?
> 
> You would need to convince gnome/plasma people to call `RebootWithFlags(1)` on logind instead of the regular one for this.
> 
> > So what is the right method to do an orderly shutdown and `reboot(kexec)` with a pre-loaded kernel?
> 
> systemd makes sure to check if a kexec kernel is already loaded before loading one itself, so `systemctl kexec` is the proper way. Maybe your systemd is too old? You could check what is happening with `SYSTEMD_LOG_LEVEL=debug systemctl --dry-run kexec`.

Indeed, the issue is fixed on v249.

**Comment by felagund** [comment]
> > Is that a systemd issue or distribution issue?
> 
> You would need to convince gnome/plasma people to call `RebootWithFlags(1)` on logind instead of the regular one for this.

Actually, turns out that KDE has it, only it is buggy, I hope they fix it soon!
https://github.com/sddm/sddm/issues/434

**Comment by iam-TJ** [comment]
Is there any reason for `systemctl kexec` to enforce the strict file-system type requirements of the BLS "just" to get the paths to the kernel and initialramfs image, and the command-line?

I can understand those being hard requirements for `sd-boot` etc., but for a `kexec` operation I'm failing to understand why simply having a BLS configuration that describes what `kexec` needs isn't allowed.

I have (UEFI and BIOS) systems all with separate LUKS protected `/boot/` ext4 file-systems that `grub2` unlocks. `vmlinuz` and `initrd.img` are within. `kexec -l /vmlinuz --initrd /initrd.img --reuse-cmdline` prepares it manually (via symbolic links in the root file-system in this case). I can create the correct BLS directory structure with a configuration file of the form:
```
title Debian 11 Bullseye
version latest
machine-id c486455029ec45a893cb6617385be0b0
architecture x64
linux /boot/vmlinuz
initrd /boot/initrd.img
options root=/dev/mapper/SUNNY-rootfs ro systemd.unified_cgroup_hierarchy=1 radeon.drm=1 initramfs.debug nosplash vt.handoff=7
```
For a `systemctl kexec` operation that should be fine, but:
```
# SYSTEMD_LOG_LEVEL=debug systemctl --dry-run kexec
Failed to check file system type of "/efi": No such file or directory
File system "/boot" is not a FAT EFI System Partition (ESP) file system.
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Found container virtualization none.
Failed to probe partition scheme of "/dev/block/253:64": Input/output error
```
As a result I have to create a separate service to do the loading:
```
# systemctl cat kexec-my-loader.service
# /etc/systemd/system/kexec-my-loader.service
[Unit]
# Loads the default kernel and initrd into memory when a reboot/restart is triggered so that
# systemd can and will do an OS reboot that skips firmware/BIOS and boot-loader stages. This is
# great for systems with an encrypted /boot/ where GRUB would normally wait for a valid key to be provided
# manually or via hardware token (e.g. Yubikey).
# Supports using "systemctl kexec" or GUI environments that use systemd-logind's DBus service
# "org.freedesktop.login1" and call "org.freedesktop.login1.Manager.RebootWithFlags(uint64 SD_LOGIND_KEXEC_REBOOT)"
# where SD_LOGIND_KEXEC_REBOOT = (1 << 1) a.k.a. 2 instead of "Manager.Reboot(bool interactive)"
Description=Load current kernel for kexec
Documentation=man:kexec(8)
DefaultDependencies=no
Before=shutdown.target umount.target final.target reboot.target systemd-reboot.service

[Service]
Type=oneshot
# these paths should point to the default (and usually last-installed) kernel and initrd.img pair
ExecStart=/usr/sbin/kexec -d -l /vmlinuz --initrd /initrd.img --reuse-cmdline

[Install]
WantedBy=reboot.target
```
and also over-ride the default `systemd-reboot.service`:
```
# systemctl cat systemd-reboot.service
# /lib/systemd/system/systemd-reboot.service
#  SPDX-License-Identifier: LGPL-2.1-or-later
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=System Reboot
Documentation=man:systemd-reboot.service(8)
DefaultDependencies=no
Requires=shutdown.target umount.target final.target
After=shutdown.target umount.target final.target
SuccessAction=reboot-force

# /etc/systemd/system/systemd-reboot.service.d/override.conf
[Unit]
SuccessAction=none

[Service]
Type=oneshot
# clear any existing commands just-in-case (usually there are none)
ExecStart=
ExecStart=/usr/bin/systemctl --force kexec
```

**Comment by keszybz** [comment]
The man page didn't describe that kexec also loads the kernel; that is tracked in #24588. The second source of confusion was that Fedora has something that *looks* like the boot loader specification setup, but actually is not. Lennart explained this above, we don't need to beat this dead horse. The other issues are with how we detect `/boot` partition and such. If there are problems in that area, they are not specific to kexec, but rather `bootctl`. Because of the inital confusion, this issue has turned into a grab-bag of loosely related things. I'll submit a patch to improve the docs. If the detection of `/boot` has shortcomings, please retest with recent versions of the code and open a new issue.
